### PR TITLE
create CUDA cache (for JIT compiled PTX code) in build dir instead of $HOME

### PR DESCRIPTION
**Pull request by Flamefire**
Add option to control the cache size and to disable it

### CODE REVIEW

File: easybuild/framework/easyblock.py [hunk 1]
```diff
@@ -2150,6 +2150,22 @@ def prepare_step(self, start_dir=True, load_tc_deps_modules=True):
             self.log.info("Loading extra modules: %s", extra_modules)
             self.modules_tool.load(extra_modules)
 
+        # Setup CUDA cache if required. If we don't do this, CUDA will use the $HOME for its cache files
+        if get_software_root('CUDA') or get_software_root('CUDAcore'):
```
Comment (boegel) [inline easybuild/framework/easyblock.py:2154]: @Flamefire This belongs in a separate function imho, which we just call here.

How about a `set_up_cuda_cache` function in filetools?

File: easybuild/tools/options.py [hunk 2]
```diff
@@ -357,6 +357,9 @@ def override_options(self):
             'consider-archived-easyconfigs': ("Also consider archived easyconfigs", None, 'store_true', False),
             'containerize': ("Generate container recipe/image", None, 'store_true', False, 'C'),
             'copy-ec': ("Copy specified easyconfig(s) to specified location", None, 'store_true', False),
+            'cuda-cache-maxsize': ("Maximum size of the CUDA cache (in bytes) used for JIT compilation of PTX code. "
```
Comment (boegel) [inline easybuild/tools/options.py:360]: Would a size in MB make more sense?

I'd say it's very unlikely that people would want to use a cache size smaller than 1MB?

File: easybuild/framework/easyblock.py [hunk 3]
```diff
@@ -2150,6 +2150,22 @@ def prepare_step(self, start_dir=True, load_tc_deps_modules=True):
...
+            cuda_cache_maxsize = build_option('cuda_cache_maxsize')
+            if cuda_cache_maxsize is None:
+                cuda_cache_maxsize = 1 * 1024 * 1024 * 1024  # 1 GB default value
+            if cuda_cache_maxsize == 0:
+                self.log.info('Disabling CUDA PTX cache as per request')
+                env.setvar('CUDA_CACHE_DISABLE', '1')
+            else:
+                cuda_cache_dir = os.path.join(self.builddir, 'eb-cuda-cache')
```
Comment (boegel) [inline easybuild/framework/easyblock.py:2162]: Don't we want to make this configurable too?

I can see myself using `/dev/shm` for this for example, but always using `/dev/shm` for `--buildpath` is probably not a good idea...

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/framework/easyblock.py (hunk 3 above)
Comment (Flamefire) [reply easybuild/framework/easyblock.py]: Added --cuda-cache-dir

File: easybuild/tools/options.py (hunk 2 above)
Comment (Flamefire) [reply easybuild/tools/options.py]: Ok done, changed this to be MiB

File: easybuild/framework/easyblock.py (hunk 1 above)
Comment (Flamefire) [reply easybuild/framework/easyblock.py:2172]: Locations considered for that function:
- filetools: This is not reading/writing of files but simply changing some env vars
- environment: That contains only the most basic functions
- toolchain: Likely very good fit but excludes allow_system_deps and extra_modules options if done in `prepare` and needs access to EasyBlock.builddir for the default
- EasyBlock: Easiest and also makes sense. Could make it private if we want to decide on the exact location later

**Review by boegel — APPROVED** [review]

**Pull request merged**